src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
+{
+    internal abstract class VirtualDocumentFactoryBase : VirtualDocumentFactory
+    {
+        private readonly ITextBufferFactoryService _textBufferFactory;
+        private readonly ITextDocumentFactoryService _textDocumentFactory;
+        private readonly FileUriProvider _fileUriProvider;
+
+        protected IContentTypeRegistryService ContentTypeRegistry { get; }
+
+        public VirtualDocumentFactoryBase(
+            IContentTypeRegistryService contentTypeRegistry,
+            ITextBufferFactoryService textBufferFactory,
+            ITextDocumentFactoryService textDocumentFactory,
+            FileUriProvider filePathProvider)
+        {
+            if (contentTypeRegistry is null)
+            {
+                throw new ArgumentNullException(nameof(contentTypeRegistry));
+            }
+
+            if (textBufferFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactory));
+            }
+
+            if (textDocumentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textDocumentFactory));
+            }
+
+            if (filePathProvider is null)
+            {
+                throw new ArgumentNullException(nameof(filePathProvider));
+            }
+
+            ContentTypeRegistry = contentTypeRegistry;
+
+            _textBufferFactory = textBufferFactory;
+            _textDocumentFactory = textDocumentFactory;
+            _fileUriProvider = filePathProvider;
+        }
+
+        protected abstract IContentType LanguageLSPContentType { get; }
+
+        /// <summary>
+        /// Converts registered document content type to one that extends "code-languageserver-base" and
+        /// CodeRemoteContentDefinition.CodeRemoteContentTypeName if needed.
+        /// </summary>
+        /// <param name="contentType">Content type registered with IContentTypeRegistryService</param>
+        /// <returns>Corresponding LSP/remote content type</returns>
+        protected virtual IContentType ConvertToLSPContentType(IContentType contentType) => contentType;
```

Comment:
Can remove this

Reply:
Oh yeah, missed it, thanks!

---
In reply to: [452549958](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r452549958) [](ancestors = 452549958)